File: backend/jasprchain/consensus/proposer.py
"""Proposer Selection using VRF
Mirrors Rust: vrf crate pattern

Implements weighted VRF rotation for fair block proposer selection
"""
from dataclasses import dataclass
from typing import List, Optional, Tuple
import hashlib
import secrets
from .validator import ValidatorSet, Validator


@dataclass
class VRFOutput:
    """Verifiable Random Function output
    
    Used for unpredictable proposer selection
    Mirrors Rust VRF implementation
    """
    value: bytes  # 32 bytes random value
    proof: bytes  # VRF proof for verification
    
    @property
    def value_hex(self) -> str:
        return self.value.hex()
    
    def as_int(self) -> int:
        """Convert to integer for weighted selection"""
        return int.from_bytes(self.value[:8], 'big')


class ProposerSelection:
    """Implements weighted VRF proposer selection
    
    Key features (from JasprChain spec):
    - Weighted by stake + performance scoring
    - Reduces predictability and censorship vulnerability
    - Fast failover for latency spikes
    """
    
    def __init__(self, validator_set: ValidatorSet):
        self.validator_set = validator_set
        self._last_proposer: Optional[str] = None
        self._round_robin_index = 0
    
    def generate_vrf(self, seed: bytes) -> VRFOutput:
        """Generate VRF output from seed
        
        In production Rust: use actual VRF (e.g., ecvrf crate)
        Simulated here with HMAC-based approach
        """
        # Seed typically includes: previous_block_hash + height + epoch
        random_key = secrets.token_bytes(32)
        
        import hmac
        value = hmac.new(random_key, seed, hashlib.sha256).digest()
        proof = hmac.new(seed, random_key, hashlib.sha256).digest()
        
        return VRFOutput(value=value, proof=proof)
    
    def select_proposer(
        self, 
        height: int, 
        previous_hash: str,
        use_vrf: bool = True
    ) -> Optional[Validator]:
        """Select proposer for a given height
        
        Selection algorithm:
        1. Generate VRF from previous block hash + height
        2. Weight by stake amount
        3. Apply performance scoring (uptime bonus)
        4. Select deterministically based on VRF output
        """
        active_validators = self.validator_set.get_active_validators()
        
        if not active_validators:
            return None
        
        if use_vrf:
            # VRF-based weighted selection
            seed = f"{previous_hash}:{height}".encode()
            vrf = self.generate_vrf(seed)
            
            # Calculate weighted scores
            scores = []
            for v in active_validators:
                # Base score from stake
                base_score = v.stake
                
                # Performance bonus (up to 10% boost for good uptime)
                uptime_bonus = int(base_score * 0.1 * (v.stats.uptime_percentage / 100))
                
                # Penalty for recent proposals (spread load)
                recent_penalty = 0
                if v.address == self._last_proposer:
                    recent_penalty = int(base_score * 0.05)
                
                final_score = base_score + uptime_bonus - recent_penalty
                scores.append((v, final_score))
            
            # Weighted selection using VRF
            total_score = sum(s[1] for s in scores)
            if total_score == 0:
                return active_validators[0]
            
            target = vrf.as_int() % total_score
            cumulative = 0
            
            for validator, score in scores:
                cumulative += score
                if cumulative > target:
                    self._last_proposer = validator.address
                    return validator
            
            return scores[-1][0]
        else:
            # Simple round-robin (fallback)
            self._round_robin_index = height % len(active_validators)
            return active_validators[self._round_robin_index]
    
    def verify_proposer(
        self, 
        proposer: str, 
        height: int, 
        previous_hash: str
    ) -> bool:
        """Verify that a proposer was validly selected
        
        Other validators use this to verify block proposals
        """
        expected = self.select_proposer(height, previous_hash)
        return expected is not None and expected.address == proposer

File: rust-core/jasprchain/src/consensus/proposer.rs
//! VRF-based proposer selection

use crate::consensus::{Validator, ValidatorSet};
use rand::Rng;

/// VRF output for proposer selection
pub struct VRFOutput {
    pub value: [u8; 32],
    pub proof: [u8; 32],
}

impl VRFOutput {
    pub fn as_int(&self) -> u64 {
        u64::from_be_bytes(self.value[..8].try_into().unwrap_or_default())
    }
}

/// Weighted VRF proposer selection
pub struct ProposerSelection<'a> {
    validator_set: &'a ValidatorSet,
    last_proposer: Option<String>,
}

impl<'a> ProposerSelection<'a> {
    pub fn new(validator_set: &'a ValidatorSet) -> Self {
        Self {
            validator_set,
            last_proposer: None,
        }
    }
    
    /// Generate VRF from seed
    pub fn generate_vrf(&self, seed: &[u8]) -> VRFOutput {
        use sha2::{Sha256, Digest};
        
        let mut rng = rand::thread_rng();
        let random_key: [u8; 32] = rng.gen();
        
        let mut hasher = Sha256::new();
        hasher.update(&random_key);
        hasher.update(seed);
        let value: [u8; 32] = hasher.finalize().into();
        
        let mut hasher = Sha256::new();
        hasher.update(seed);
        hasher.update(&random_key);
        let proof: [u8; 32] = hasher.finalize().into();
        
        VRFOutput { value, proof }
    }
    
    /// Select proposer for a given height
    pub fn select_proposer(&mut self, height: u64, previous_hash: &str) -> Option<&Validator> {
        let active = self.validator_set.get_active_validators();
        if active.is_empty() {
            return None;
        }
        
        let seed = format!("{}:{}", previous_hash, height);
        let vrf = self.generate_vrf(seed.as_bytes());
        
        // Calculate weighted scores
        let mut scores: Vec<(&Validator, u64)> = active
            .iter()
            .map(|v| {
                let base = v.stake;
                let uptime_bonus = (base as f64 * 0.1 * (v.stats.uptime_percentage / 100.0)) as u64;
                let penalty = if Some(v.address.clone()) == self.last_proposer {
                    (base as f64 * 0.05) as u64
                } else {
                    0
                };
                (*v, base + uptime_bonus - penalty)
            })
            .collect();
        
        let total_score: u64 = scores.iter().map(|(_, s)| s).sum();
        if total_score == 0 {
            return Some(active[0]);
        }
        
        let target = vrf.as_int() % total_score;
        let mut cumulative = 0u64;
        
        for (validator, score) in &scores {
            cumulative += score;
            if cumulative > target {
                self.last_proposer = Some(validator.address.clone());
                return Some(validator);
            }
        }
        
        scores.last().map(|(v, _)| *v)
    }
}
